concordium-consensus/src/Concordium/GlobalState/Statistics.hs

```diff
@@ -74,3 +78,151 @@ initialConsensusStatistics =
           _finalizationPeriodEMA = Nothing,
           _finalizationPeriodEMVar = Nothing
         }
+
+-- |Update the statistics when a block becomes fully validated (arrives).
+updateStatsOnArrive ::
+    -- |Nominal block time.
+    UTCTime ->
+    -- |Block arrival time.
+    UTCTime ->
+    -- |Number of transactions in the block.
+    Int ->
+    ConsensusStatistics ->
+    ConsensusStatistics
+updateStatsOnArrive nominalTime arrivalTime transactionCount =
+    updateTransactionsPerBlock . updatePeriod . updateLatency . (blocksVerifiedCount +~ 1)
+  where
+    updateLatency s =
+        s
+            & (blockArriveLatencyEMA .~ oldEMA + emaWeight * delta)
+            & (blockArriveLatencyEMVar %~ \oldEMVar -> (1 - emaWeight) * (oldEMVar + emaWeight * delta * delta))
+      where
+        oldEMA = s ^. blockArriveLatencyEMA
+        delta = realToFrac (diffUTCTime arrivalTime nominalTime) - oldEMA
+
+    updatePeriod s =
```

Comment:
Now we're at it. Maybe we should extend the documentation a bit. For example, what is meant by "period" in this context?